README.md
# squishy

A Clojure library for safely and reliably consuming SQS messages.

## Usage

```clj
(ns demo
  [squishy.core :as sqs])

(defonce consumer-id (atom nil))

(defn process-message [message]
  (println "Processing: " message)
  ;; ... do something fancy
  )

(let [creds {:access-key "AKIA..."
             :access-secret "8jcNZ9mM..."
             :region #aws/region "US_EAST_1"}
      cid (sqs/consume-messages creds
                                "incoming-queue"
                                "failure-queue"
                                process-message)]
  (reset! consumer-id cid))

;; later when you are done:

(sqs/stop-consumer @consumer-id)
```

### Region

You can send several values in for the `:region` component of the credentials.

* "us-east-1": you can send in a string that conforms to the "name" of the region, lowercase with dashes
* "US_EAST_1": you can send in a string that conforms to the string version of the Regions enum, uppercase with underscores
* com.amazonaws.regions.Region: you can send in an actual instance of the Region class

There's a data-reader tag defined as `#aws/region` that can convert a string in either of
the top two formats into a Region class instance as well.

### Shutting down

**IMPORTANT:** Squishy is quite aggressive at recovering from exceptions and
resuming processing of messages. This behavior is necessary to maintain 
long-term reliable connections to SQS.

A side effect of this is that squishy will interpret the exceptions that it 
receives when you stop / undeploy your app as things to recover from. If you
see the "Consuming SQS messages from..." log message after attempting to stop a
squishy app, then you are running into this problem.

As of version 3.0.0, you **MUST** keep the `consumer-id` returned by 
`consume-messages` and call `(squishy.core/stop-consumer consumer-id)` when
you are shutting down / undeploying your app. This will tell squishy to stop
trying to recover that consumer.

### Visibility timeouts

> For some background on visibility timeouts in SQS, we recommend [reading 
Amazon's documentation](http://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/AboutVT.html).

In version 3.0.0, squishy gained the ability to adjust the visibility timeout
of the messages it's processing automatically. The visibility timeout default
is 30 seconds, which means that 30 seconds after a consumer starts processing a
message, SQS will make it available to other consumers if the original consumer
hasn't deleted it. Since this is not usually what you expect or want, squishy 
now informs SQS that it should increase the visibility timeout of individual 
messages when it knows that the provided processing function is still 
processing that message. This means that other consumers won't see the message.

If your SQS queue has a different visibility timeout configured, you can use
the 5-arity version of `consume-messages` to tell squishy when it should raise
the visibility timeout. So, for example, if you have configured your queue with
a 300-second visibility timeout, you should call `consume-messages` like this:

```
(sqs/consume-messages creds
                      "incoming-queue"
                      "failure-queue"
                      {:visibility-timeout 300}
                      process-message)
```

It doesn't hurt to not do this as squishy assumes 30 second visibility timeouts
by default. It would just result in more SQS API calls than are strictly
necessary.

### Very Long Running Jobs

If you sometimes/often have jobs that run over the maximum visibility timeout
for SQS (12 hours), you will get new copies of the message spawning without
bounds every 12 hours until you manually intervene. Since this is undesireable,
there is an option you can provide to `consume-messages` called `:delete-callback`.
When set to true, your processing function will receive two arguments, the
message to process and a callback function to be called when enough information has
been stored to keep track of the job. Calling this function will then delete the
message from SQS and no copies will spawn (presuming you have called this prior
to the 12 hour maximum visibility timeout).

```
(defn my-processor [message delete-callback]
  (record-job message)
  (delete-callback)
  (very-long-job-kickoff message))


(sqs/consume-messages creds q-name f-q-name {:delete-callback true} my-processor)
```

## License

Copyright © 2014-2018 Democracy Works, Inc.

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


ChangeLog.md
# Change Log

## Changes between 3.0.1 and 3.1.0

### Improved Region configuration support

Whereas before the only acceptable string format for the `#aws/region` tag
was the region enum format (i.e. "US_WEST_1"), the tag can now also take the
more common name format (i.e. "us-west-1"). The `start-consumer` now can take
either of the string formats or a `com.amazonaws.regions.Region` instance, whereas
before it only took the instance type.

See the Region section of the README.

## Changes between 3.0.0 and 3.0.1

### Excluded `logback.xml` from library JAR file

Logging config files don't belong in library JARs, and having this one in there
resulted in logback-using consumer code generating warnings about multiple
logback configs on the classpath. Library consumers should get full control
over their logging configs, warning-free.

## Changes between 2.0.0 and 3.0.0

### `consume-messages` now returns a consumer-id instead of a future

In 1.0.0 it returned a future that you could cancel when you wanted to stop
consuming messages. But in 2.0.0 that stopped working because of the retry
loop added in that version.

In 3.0.0 it will return a consumer-id instead. You can then call
`(squishy.core/stop-consumer consumer-id)` when you want to stop processing.

**This is a breaking change.**

### `consume-messages` will now handle increasing the visibility timeout of SQS messages

By default SQS queues have a visibility timeout of 30 seconds (though you can
increase it in each queue's config). This means that if it takes your code
longer than that timeout to process a message, SQS will make it available to
other consumers. In 3.0.0, squishy will now tell SQS to increase the visibility
timeout of a message that is still being processed so that you don't have
redundant processing.

## Changes between Squishy 1.0.0 and 2.0.0

### `consume-messages` can recover from some SQS connection problems

`consume-messages` used to not be able to recover from, for example, a
connection reset when trying to read a message from the queue. And
since `consume-messages` does its work in a separate thread, the
exception thrown would not be handled. Now, however, Squishy will
attempt to retry, backing off, before finally giving up, logging a
fatal error, and realizing the future.

This required the first argument for `consume-messages` to
change. Instead of passing a client, you provide a map of
configuration containing your `:access-key`, `:access-secret` and
`:region`. Squishy will create clients as needed.

**This is a breaking change.**


test/squishy/core_test.clj
(ns squishy.core-test
  (:require [clojure.test :refer :all]
            [squishy.core :refer :all]
            [cemerick.bandalore :as sqs]))

(defn blocking-consume-messages [f]
  (let [consumer-id (consume-messages nil nil nil f)]
    @(get @consumer-futures consumer-id)))

(deftest consume-messages-test
  (testing "calls function arg on each msg"
    (let [result (atom [])]
      (with-redefs [client (constantly nil)
                    sqs/create-queue (constantly nil)
                    sqs/delete (constantly nil)
                    sqs/change-message-visibility (constantly nil)
                    sqs/polling-receive (constantly ["msg 1" "msg 2"])]
        (blocking-consume-messages #(swap! result conj
                                           (clojure.string/upper-case %)))
        (is (= ["MSG 1" "MSG 2"]
               @result)))))

  (testing "deletes msgs as it consumes them"
    (let [queue (atom ["msg 1" "msg 2"])]
      (with-redefs [client (constantly nil)
                    sqs/create-queue (constantly nil)
                    sqs/change-message-visibility (constantly nil)
                    sqs/polling-receive (constantly @queue)
                    sqs/delete (fn [client message]
                                 (swap! queue #(drop 1 %)))]
        (blocking-consume-messages (constantly nil))
        (is (= []
               (do
                 (blocking-consume-messages (constantly nil))
                 @queue))))))

  (testing "puts msgs on fail queue when processing fn throws exception"
    (let [fail-queue (atom [])]
      (with-redefs [client (constantly nil)
                    sqs/create-queue (constantly nil)
                    sqs/change-message-visibility (constantly nil)
                    sqs/polling-receive (constantly [{:body "msg 1"}
                                                     {:body "msg 2"}])
                    sqs/delete (constantly nil)
                    sqs/send (fn [client q msg] (swap! fail-queue conj msg))]
        (blocking-consume-messages
         #(throw (Exception. (str (:body %) " failed"))))
        (is (= ["{:body \"msg 1\", :error \"msg 1 failed\"}"
                "{:body \"msg 2\", :error \"msg 2 failed\"}"]
               @fail-queue)))))

  (testing "can recover from polling-receive errors"
    (let [queue (atom (list (constantly 1)
                            (fn [& _] (throw (RuntimeException. "1")))
                            (constantly 2)
                            (fn [& _] (throw (RuntimeException. "2")))
                            (constantly 3)))
          done (atom [])]
      (with-redefs [client (constantly nil)
                    sqs/create-queue (constantly nil)
                    sqs/polling-receive (fn [& _]
                                          (letfn [(r []
                                                    (lazy-seq
                                                     (when (seq @queue)
                                                       (let [x (first @queue)]
                                                         (swap! queue rest)
                                                         (cons (x) (r))))))]
                                            (r)))
                    sqs/delete (constantly nil)
                    sqs/send (constantly nil)]
        (blocking-consume-messages #(swap! done conj %))
        (is (= [1 2 3]
               @done)))))

  (testing "fails after n times"
    (let [queue (atom (concat
                       (map constantly (range 3))
                       (map #(fn [& _]
                               (throw (RuntimeException. (str %))))
                            (range 10))
                       (map constantly (range 3))))
          done (atom [])]
      (with-redefs [client (constantly nil)
                    backoff-start 1 ;; so it doesn't take forever
                    max-failures 10
                    sqs/create-queue (constantly nil)
                    sqs/polling-receive (fn [& _]
                                          (letfn [(r []
                                                    (lazy-seq
                                                     (when (seq @queue)
                                                       (let [x (first @queue)]
                                                         (swap! queue rest)
                                                         (cons (x) (r))))))]
                                            (r)))
                    sqs/delete (constantly nil)
                    sqs/send (constantly nil)]
        (blocking-consume-messages #(swap! done conj %))
        (is (= [0 1 2]
               @done)))))

  (testing "resets after every success"
    (let [queue (atom (interleave (map constantly (range 100))
                                  (map #(fn [& _]
                                          (throw (RuntimeException. (str %))))
                                       (range 100))))
          done (atom [])]
      (with-redefs [client (constantly nil)
                    backoff-start 1 ;; so it doesn't take forever
                    max-failures 10
                    sqs/create-queue (constantly nil)
                    sqs/polling-receive (fn [& _]
                                          (letfn [(r []
                                                    (lazy-seq
                                                     (when (seq @queue)
                                                       (let [x (first @queue)]
                                                         (swap! queue rest)
                                                         (cons (x) (r))))))]
                                            (r)))
                    sqs/delete (constantly nil)
                    sqs/send (constantly nil)]
        (blocking-consume-messages #(swap! done conj %))
        (is (= (range 100)
               @done))))))


test/democracyworks/squishy_test.clj
(ns democracyworks.squishy-test
  (:require [clojure.test :refer :all]
            [democracyworks.squishy :refer :all :as squishy]
            [cemerick.bandalore :as sqs]))

(deftest consume-messages-test
  (testing "calls function arg on each msg"
    (let [result (atom [])]
      (with-redefs [squishy/get-queue (constantly nil)
                    sqs/delete (constantly nil)
                    sqs/change-message-visibility (constantly nil)
                    sqs/polling-receive (constantly ["msg 1" "msg 2"])]
        (is (= ["MSG 1" "MSG 2"]
               (do
                 @(consume-messages nil
                                    #(swap! result conj
                                            (clojure.string/upper-case %)))
                 @result))))))

  (testing "deletes msgs as it consumes them"
    (let [queue (atom ["msg 1" "msg 2"])]
      (with-redefs [squishy/get-queue (constantly nil)
                    sqs/change-message-visibility (constantly nil)
                    sqs/polling-receive (constantly @queue)
                    sqs/delete (fn [client message]
                                 (swap! queue #(drop 1 %)))]
        (is (= []
               (do
                 @(consume-messages nil (constantly nil))
                 @queue))))))

  (testing "puts msgs on fail queue when processing fn throws exception"
    (let [fail-queue (atom [])]
      (with-redefs [squishy/get-queue (constantly nil)
                    squishy/get-fail-queue (constantly nil)
                    sqs/change-message-visibility (constantly nil)
                    sqs/polling-receive (constantly [{:body "msg 1"}
                                                     {:body "msg 2"}])
                    sqs/delete (constantly nil)
                    sqs/send (fn [client q msg] (swap! fail-queue conj msg))]
        (is (= ["{:body \"msg 1\", :error \"msg 1 failed\"}"
                "{:body \"msg 2\", :error \"msg 2 failed\"}"]
               (do @(consume-messages
                     nil #(throw (Exception. (str (:body %) " failed"))))
                   @fail-queue)))))))


src/democracyworks/squishy.clj
(ns democracyworks.squishy
  (:require [democracyworks.squishy.data-readers]
            [cemerick.bandalore :as sqs]
            [turbovote.resource-config :refer [config]]
            [clojure.tools.logging :refer [error info]]))

(defn client []
  (doto (sqs/create-client (config :aws :creds :access-key)
                           (config :aws :creds :secret-key))
    (.setRegion (config :aws :sqs :region))))

(defn- create-queue [client queue-key]
  (sqs/create-queue client (config :aws :sqs queue-key)))

(def ^:private memoized-create-queue (memoize create-queue))

(defn- get-queue [client]
  (memoized-create-queue client :queue))

(defn- get-fail-queue [client]
  (memoized-create-queue client :fail-queue))

(defn- report-error [client body error]
  (let [q (get-fail-queue client)]
    (sqs/send client q (pr-str {:body body :error (.getMessage error)}))))

(defn- safe-process [client f]
  (fn [message]
    (info "Processing SQS message:" (str "<<" message ">>"))
    (let [q (get-queue client)
          processor (future
                      (try (f message)
                           (catch Exception e
                             (let [body (:body message)]
                               (error e "Failed to process" body)
                               (report-error client body e)))))]
      (loop [timeout 30] ; default visibility timeout is 30 secs
        (if (realized? processor)
          @processor
          (let [new-timeout (int (* timeout 1.5))]
            (Thread/sleep (* (- timeout 10) 1000))
            (sqs/change-message-visibility client q message new-timeout)
            (recur new-timeout)))))))

(defn consume-messages
  [client f]
  (let [q (get-queue client)]
    (future
      (do
        (info "Consuming SQS messages from" q)
        (dorun
         (map (sqs/deleting-consumer client (safe-process client f))
              (sqs/polling-receive client q :max-wait Long/MAX_VALUE :limit 10)))))))


src/squishy/core.clj
(ns squishy.core
  (:require [squishy.data-readers :as dr]
            [cemerick.bandalore :as sqs]
            [clojure.tools.logging :as log])
  (:import  [com.amazonaws.regions Region]))

(def default-visibility-timeout 30) ; seconds

(defmulti client
  "Construct an SQS client instance from the provided credentials and region.
   Region can be a string in either name (`us-east-1`) or enum (`US_EAST_1`)
   format, or it can be an instance of com.amazonaws.regions.Region already."
  (fn [_ _ region] (type region)))

(defmethod client com.amazonaws.regions.Region [access-key secret-key region]
  (doto (sqs/create-client access-key secret-key)
    (.setRegion region)))

(defmethod client java.lang.String [access-key secret-key region]
  (client access-key secret-key (dr/aws-region region)))

(defmethod client :default [_ _ region]
  (throw (RuntimeException. "Region must either be a string in either \"us-west-1\" or \"US_WEST_1\" format or an instance of com.amazonaws.regions.Region. You can also use the `#aws/region` data-reader tag in a configuration to convert a string in the above formats to the Java instance.")))

(defn report-error [client fail-queue-url body error]
  (let [fail-message (pr-str {:body body :error (.getMessage error)})]
    (sqs/send client fail-queue-url fail-message)))

(defn safe-process
  [client queue-url fail-queue-url visibility-timeout delete-callback f]
  (fn [message]
    (log/debug "Processing SQS message:" (pr-str message))
    (let [timeout-increaser (future
                              (loop [timeout visibility-timeout]
                                (let [new-timeout (int (* timeout 1.5))]
                                  (Thread/sleep (* 1000 timeout 0.75))
                                  (sqs/change-message-visibility client queue-url message new-timeout)
                                  (recur new-timeout))))
          callback (when delete-callback
                     (fn []
                       (sqs/delete client message)
                       (future-cancel timeout-increaser)))]
      (try (if callback
             (f message callback)
             (f message))
           (catch Exception e
             (let [body (:body message)]
               (log/error e "Failed to process" body)
               (report-error client fail-queue-url body e)))
           (finally
             (future-cancel timeout-increaser))))))

(def backoff-start 1000)
(def max-failures  2)

(defonce consumer-futures (atom {}))

(defn with-backoff* [backoff-start max-failures consumer-id f]
  (let [rec (atom 0)
        reset (fn [] (reset! rec 0))]
    (loop []
      (when (try
              (f reset)
              (swap! consumer-futures dissoc consumer-id)
              nil
              (catch Exception t
                (when (get @consumer-futures consumer-id)
                  (log/error "Failure:" t)
                  (swap! rec inc)
                  :keep-trying)))
        (if (>= @rec max-failures)
          (log/fatal "Failed" @rec "times, exiting.")
          (do
            (Thread/sleep (* backoff-start (Math/pow 2 (dec @rec))))
            (when (get @consumer-futures consumer-id)
              (recur))))))))

(defmacro with-backoff [[reset backoff-start max-failures consumer-id] & body]
  `(with-backoff* ~backoff-start ~max-failures ~consumer-id
     (fn [~reset] ~@body)))

(defn consume-messages
  ([creds queue-name fail-queue-name f]
   (consume-messages creds queue-name fail-queue-name
                     {:visibility-timeout default-visibility-timeout}
                     f))
  ([creds queue-name fail-queue-name options f]
   (let [consumer-id (java.util.UUID/randomUUID)
         message-future
         (future
           (with-backoff [reset backoff-start max-failures consumer-id]
             (let [client (client (:access-key creds)
                                  (:access-secret creds)
                                  (:region creds))
                   queue-url (sqs/create-queue client queue-name)
                   fail-queue-url (sqs/create-queue client fail-queue-name)
                   visibility-timeout (get options :visibility-timeout
                                           default-visibility-timeout)
                   delete-callback (get options :delete-callback false)
                   consume (sqs/deleting-consumer
                            client
                            (safe-process client
                                          queue-url
                                          fail-queue-url
                                          visibility-timeout
                                          delete-callback
                                          f))]
               (log/info "Consuming SQS messages from" queue-url)
               (doseq [message (sqs/polling-receive client
                                                    queue-url
                                                    :max-wait Long/MAX_VALUE
                                                    :limit 10)]
                 (reset)
                 (consume message)))))]
     (swap! consumer-futures assoc consumer-id message-future)
     consumer-id)))

(defn stop-consumer [consumer-id]
  (swap! consumer-futures dissoc consumer-id))
